src/surface.cpp

```diff
@@ -983,6 +1004,212 @@ void SurfaceQuadric::to_hdf5_inner(hid_t group_id) const
   write_dataset(group_id, "coefficients", coeffs);
 }
 
+//==============================================================================
+// Torus helper functions
+//==============================================================================
+
+double torus_distance(double x1, double x2, double x3, double u1, double u2,
+  double u3, double A, double B, double C, bool coincident)
+{
+  // Coefficients for equation: (c2 t^2 + c1 t + c0)^2 = c2' t^2 + c1' t + c0'
+  double D = (C * C) / (B * B);
+  double c2 = u1 * u1 + u2 * u2 + D * u3 * u3;
+  double c1 = 2 * (u1 * x1 + u2 * x2 + D * u3 * x3);
+  double c0 = x1 * x1 + x2 * x2 + D * x3 * x3 + A * A - C * C;
+  double four_A2 = 4 * A * A;
+  double c2p = four_A2 * (u1 * u1 + u2 * u2);
+  double c1p = 2 * four_A2 * (u1 * x1 + u2 * x2);
+  double c0p = four_A2 * (x1 * x1 + x2 * x2);
+
+  // Coefficient for equation: a t^4 + b t^3 + c t^2 + d t + e = 0
+  double coeff[5];
+  coeff[0] = c0 * c0 - c0p;
+  coeff[1] = 2 * c0 * c1 - c1p;
+  coeff[2] = c1 * c1 + 2 * c0 * c2 - c2p;
+  coeff[3] = 2 * c1 * c2;
+  coeff[4] = c2 * c2;
+
+  std::complex<double> roots[4];
+  oqs::quartic_solver(coeff, roots);
+
+  // Find smallest positive, real root. In the case where the particle is
+  // coincident with the surface, we are sure to have one root very close to
+  // zero but possibly small and positive. A tolerance is set to discard that
+  // zero.
+  double distance = INFTY;
+  double cutoff = coincident ? 1e-9 : 0.0;
+  for (int i = 0; i < 4; ++i) {
+    if (roots[i].imag() == 0) {
+      double root = roots[i].real();
+      if (root > cutoff && root < distance) {
+        distance = root;
+      }
+    }
+  }
+  return distance;
+}
+
+//==============================================================================
+// SurfaceXTorus implementation
+//==============================================================================
+
+SurfaceXTorus::SurfaceXTorus(pugi::xml_node surf_node) : CSGSurface(surf_node)
+{
+  read_coeffs(surf_node, id_, x0_, y0_, z0_, A_, B_, C_);
+}
+
+void SurfaceXTorus::to_hdf5_inner(hid_t group_id) const
+{
+  write_string(group_id, "type", "x-torus", false);
+  std::array<double, 6> coeffs {{x0_, y0_, z0_, A_, B_, C_}};
+  write_dataset(group_id, "coefficients", coeffs);
+}
+
+// todo should do some alebgra first to get rid of the sqrt
```

Comment:
Could we place this comment closer to the relevant `sqrt`?

Reply:
I'm just going to remove this comment -- on modern architectures `sqrt` is actually about as expensive as a `div` instruction, so probably not something we need to be overly worried about